controllers/volumereplicationgroup_controller.go

```diff
@@ -409,94 +420,229 @@ func (v *VRGInstance) processAsPrimary() (ctrl.Result, error) {
+		if err = v.createVR(vrNamespacedName); err != nil {
+			log.Error(err, "Failed to create VolumeReplication resource", "resource", vrNamespacedName)
 
-			if err = v.createVolumeReplicationForPVC(pvc); err == nil {
-				continue
-			}
+			return fmt.Errorf("failed to create VolumeReplication resource"+
+				" (%s/%s) belonging to VolumeReplicationGroup (%s/%s), %w",
+				vrNamespacedName.Name, vrNamespacedName.Namespace, v.instance.Name, v.instance.Namespace, err)
 		}
+	}
+
+	if volRep.Spec.State == volrep.ReplicationPrimary {
+		return nil
+	}
 
-		log.Info("Requeuing due to failure in getting or creating VolumeReplication resource for PersistentVolumeClaim",
-			"errorValue", err)
+	volRep.Spec.State = volrep.ReplicationPrimary
+	if err = v.reconciler.Update(v.ctx, volRep); err != nil {
+		log.Error(err, "Failed to update VolumeReplication resource",
+			"resource", vrNamespacedName,
+			"state", volrep.ReplicationPrimary)
 
-		// requeue on failure to ensure any PVC not having a corresponding VR CR
-		requeue = true
+		return fmt.Errorf("failed to update VolumeReplication resource"+
+			" (%s/%s) as %s, belonging to VolumeReplicationGroup (%s/%s), %w",
+			vrNamespacedName.Name, vrNamespacedName.Namespace, volrep.ReplicationPrimary,
+			v.instance.Name, v.instance.Namespace, err)
 	}
 
-	return requeue
+	return nil
+}
+
+// createVolumeReplicationCRForPVC creates a VolumeReplication CR with a PVC as its data source.
+func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName) error {
+	cr := &volrep.VolumeReplication{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      vrNamespacedName.Name,
+			Namespace: vrNamespacedName.Namespace,
+		},
+		Spec: volrep.VolumeReplicationSpec{
+			DataSource: &corev1.TypedLocalObjectReference{
+				Kind: "PersistentVolumeClaim",
+				Name: vrNamespacedName.Name,
+			},
+			// Get the state of VolumeReplication from
+			// VolumeReplicationGroupSpec
+			State: v.instance.Spec.ReplicationState,
+		},
+	}
+
+	return v.reconciler.Create(v.ctx, cr)
+}
+
+// processPVCAsVRProtected processes a given PVC that is VR protected, to ensure post protection
+// tasks are complete for the given PVC
+func (v *VRGInstance) processPVCAsVRProtected(pvc *corev1.PersistentVolumeClaim,
+	log logr.Logger) bool {
+	const requeue bool = true
+
+	// Check if PVC is already protected
+	pvcProtected, err := v.ensurePVCVRProtected(pvc, log)
+	if err != nil {
+		log.Info("Requeuing as determining PersistentVolumeClaim protection failed", "errorValue", err)
+
+		return requeue
+	}
+
+	if pvcProtected {
+		return !requeue
+	}
+
+	// Backup PV bound to PVC, before creating VR for PVC
+	log.Info("Uploading PersistentVolume")
+
+	// If faulted here, PV will be present but VR may not be enabled causing application failure on peer as,
+	// PVC on peer will bind to PV, and in reality the backend will not have any volume as VR is not yet created.
+	// Hence, we should create VR and then backup PV once VR status is as expected
```

Comment:
If order is to wait for appropriate VR status and then upload PV resource, upon failover prior to upload of PV, a new PV will be created on target cluster possibly resulting in silent data loss, which I believe is OK for Regional DR (as the user has agreed to data loss) but may not be OK for Metro DR.  Whereas, if the order is upload PV first and then creat VR, the error to bind the PV to ceph volume may be desirable in Metro DR rather than silent data loss.  My thought was... what is the point of data replication if PV metadata can't be replicated, with the assumption that PV metadata is required to find and use the replicated PV data.  Can AVR use managed cluster view to query ceph volumes before it restores the PV on target cluster?  Perhaps you proposed that AVR could throw an error during failover if it can't find PV metadata in the S3 store for each PVC of the App?

Irrespective of the order, VRG status for the subscription should report OK only after VR is in the appropriate state and the PV resource has been uploaded.

Reply:
Current change of code, uploads PV before creating VR. The justification I finally took was not to fail (to preserve data) silently. The corner case where this happens will be detected and (an instance of) the application may not come online during failover.